File: api-go/internal/middlewares/rate_limiter.go
package middlewares

import (
	"fmt"
	"net/http"
	"sync"
	"time"

	e "github.com/cymoo/mote/internal/errors"
)

type windowEntry struct {
	count   int64
	resetAt time.Time
}

// RateLimit returns a net/http middleware that enforces rate limiting using an in-memory counter.
// expires: duration for rate limit window
// maxCount: maximum number of requests allowed within the window
func RateLimit(expires time.Duration, maxCount int64) func(http.Handler) http.Handler {
	var mu sync.Mutex
	store := make(map[string]*windowEntry)

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			key := fmt.Sprintf("rate:%s", r.URL.Path)

			if !checkRateLimit(&mu, store, key, expires, maxCount) {
				e.SendJSONError(w, http.StatusTooManyRequests, "too_many_attempts")
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

// checkRateLimit checks and increments the in-memory counter for the given key.
// Returns true if the request is within the limit.
func checkRateLimit(mu *sync.Mutex, store map[string]*windowEntry, key string, expires time.Duration, maxCount int64) bool {
	mu.Lock()
	defer mu.Unlock()

	now := time.Now()
	entry, ok := store[key]
	if !ok || !now.Before(entry.resetAt) {
		store[key] = &windowEntry{count: 1, resetAt: now.Add(expires)}
		return true
	}

	if entry.count >= maxCount {
		return false
	}
	entry.count++
	return true
}

File: api-rs/src/middleware/rate_limiter.rs
use crate::errors::ApiError::TooManyRequests;
use crate::errors::ApiResult;
use axum::extract::Request;
use axum::middleware::Next;
use axum::response::Response;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

/// (count, reset_at)
pub type RateLimitState = Arc<Mutex<HashMap<String, (u64, Instant)>>>;

pub fn new_state() -> RateLimitState {
    Arc::new(Mutex::new(HashMap::new()))
}

/// Middleware function to enforce rate limiting for incoming requests using an in-memory counter.
///
/// # Arguments
/// * `state` - Shared in-memory counter map.
/// * `expires` - The expiration duration for the rate limit window.
/// * `max_count` - The maximum number of requests allowed within the time window.
/// * `req` - The incoming HTTP request.
/// * `next` - The next middleware or handler in the chain.
pub async fn rate_limit(
    state: RateLimitState,
    expires: Duration,
    max_count: u64,
    req: Request,
    next: Next,
) -> ApiResult<Response> {
    let key = format!("rate:{}", req.uri().path());

    let allowed = {
        let mut store = state.lock().expect("rate limit mutex poisoned");
        let now = Instant::now();
        match store.get_mut(&key) {
            Some((count, reset_at)) if now < *reset_at => {
                if *count >= max_count {
                    false
                } else {
                    *count += 1;
                    true
                }
            }
            _ => {
                store.insert(key, (1, now + expires));
                true
            }
        }
    };

    if !allowed {
        return Err(TooManyRequests("Too many attempts, try again later".to_owned()));
    }

    Ok(next.run(req).await)
}
